pkg/container/types/tuple.go

```diff
@@ -0,0 +1,663 @@
+	p.buf = append(p.buf, b)
+}
+
+func (p *packer) putBytes(b []byte) {
+	p.buf = append(p.buf, b...)
+}
+
+func (p *packer) putBytesNil(b []byte, i int) {
+	for i >= 0 {
+		p.putBytes(b[:i+1])
+		p.putByte(0xFF)
+		b = b[i+1:]
+		i = bytes.IndexByte(b, 0x00)
+	}
+	p.putBytes(b)
+}
+
+func (p *packer) encodeBytes(code byte, b []byte) {
+	p.putByte(code)
+	if i := bytes.IndexByte(b, 0x00); i >= 0 {
+		p.putBytesNil(b, i)
+	} else {
+		p.putBytes(b)
+	}
+	p.putByte(0x00)
+}
+
+func (p *packer) encodeUint(i uint64) {
+	if i == 0 {
+		p.putByte(intZeroCode)
+		return
+	}
+
+	n := bisectLeft(i)
+	var scratch [8]byte
+
+	p.putByte(byte(intZeroCode + n))
+	binary.BigEndian.PutUint64(scratch[:], i)
+
+	p.putBytes(scratch[8-n:])
+}
+
+func (p *packer) encodeInt(i int64) {
+	if i >= 0 {
+		p.encodeUint(uint64(i))
+		return
+	}
+
+	n := bisectLeft(uint64(-i))
+	var scratch [8]byte
+
+	p.putByte(byte(intZeroCode - n))
+	offsetEncoded := int64(sizeLimits[n]) + i
+	binary.BigEndian.PutUint64(scratch[:], uint64(offsetEncoded))
+
+	p.putBytes(scratch[8-n:])
+}
+
+func (p *packer) encodeFloat32(f float32) {
+	var scratch [4]byte
+	binary.BigEndian.PutUint32(scratch[:], math.Float32bits(f))
+	adjustFloatBytes(scratch[:], true)
+
+	p.putByte(float32Code)
+	p.putBytes(scratch[:])
+}
+
+func (p *packer) encodeFloat64(d float64) {
+	var scratch [8]byte
+	binary.BigEndian.PutUint64(scratch[:], math.Float64bits(d))
+	adjustFloatBytes(scratch[:], true)
+
+	p.putByte(float64Code)
+	p.putBytes(scratch[:])
+}
+
+func (p *packer) encodeTuple(t Tuple) {
+	for i, e := range t {
+		switch e := e.(type) {
+		case nil:
+			p.putByte(nilCode)
+		case int8:
+			p.putByte(int8Code)
+			p.encodeInt(int64(e))
+		case int16:
+			p.putByte(int16Code)
+			p.encodeInt(int64(e))
+		case int32:
+			p.putByte(int32Code)
+			p.encodeInt(int64(e))
+		case int64:
+			p.putByte(int64Code)
+			p.encodeInt(e)
+		case uint8:
+			p.putByte(uint8Code)
+			p.encodeUint(uint64(e))
+		case uint16:
+			p.putByte(uint16Code)
+			p.encodeUint(uint64(e))
+		case uint32:
+			p.putByte(uint32Code)
+			p.encodeUint(uint64(e))
+		case uint64:
+			p.putByte(uint64Code)
+			p.encodeUint(e)
+		case float32:
+			p.encodeFloat32(e)
+		case float64:
+			p.encodeFloat64(e)
+		case bool:
+			if e {
+				p.putByte(trueCode)
+			} else {
+				p.putByte(falseCode)
+			}
+		case Date:
+			p.putByte(dateCode)
+			p.encodeInt(int64(e))
+		case Datetime:
+			p.putByte(datetimeCode)
+			p.encodeInt(int64(e))
+		case Timestamp:
+			p.putByte(timestampCode)
+			p.encodeInt(int64(e))
+		case Decimal64:
+			p.putByte(decimal64Code)
+			b := [8]byte(e)
+			p.encodeBytes(bytesCode, b[:])
+		case Decimal128:
+			p.putByte(decimal128Code)
+			b := [16]byte(e)
+			p.encodeBytes(bytesCode, b[:])
+		case []byte:
+			p.putByte(stringTypeCode)
+			p.encodeBytes(bytesCode, e)
+		default:
+			panic(fmt.Sprintf("unencodable element at index %d (%v, type %T)", i, t[i], t[i]))
+		}
+	}
+}
+
+func (t Tuple) Pack() []byte {
```

Comment:
Two unnecessary overheads, the first for conversion to tuples and the second for per-line processing.